anilist/src/test/java/com/huchihaitachi/anilist/AnilistPresenterTest.kt

```diff
@@ -1,80 +1,342 @@
-      anilistView.render(expectedResult)
+    setupTrampoline()
+    val initialState = AnilistViewState(pageState = PageState(null, 0, true))
+    val presenter = AnilistPresenter(
+      loadPage, refreshPage, loadAnime, getStringResource, initialState, rxSchedulers
+    )
+    val loadPageIntentMock: PublishSubject<Unit> = PublishSubject.create()
+    mockIntents(loadPageIntent = loadPageIntentMock)
+    every { loadPage(any(), any()) } returns Single.just(
+      Page(1, anime.size, true, anime, 0, 0)
+    )
+    presenter.bind(view)
+    presenter.bindIntents()
+    val expectedInitial = AnilistViewState(pageState = PageState(null, 0, true))
+    val expectedLoading = AnilistViewState(
+      loading = PAGE,
+      pageState = PageState(null, 0, true),
+    )
+    val expectedResult = AnilistViewState(pageState = PageState(anime, 1, true))
+    loadPageIntentMock.onNext(Unit)
+    verifyOrder {
+      view.render(expectedInitial)
+      view.render(expectedLoading)
+      view.render(expectedResult)
+    }
+  }
+
+  @Test
+  fun `refresh test`() {
+    setupTrampoline()
+    val initialState = AnilistViewState(pageState = PageState(anime, 1, true))
+    val presenter = AnilistPresenter(
+      loadPage, refreshPage, loadAnime, getStringResource, initialState, rxSchedulers
+    )
+    val refreshIntentMock: PublishSubject<Unit> = PublishSubject.create()
+    mockIntents(refreshIntent = refreshIntentMock)
+    val freshAnime = listOf(
+      Anime(11), Anime(21), Anime(31), Anime(41), Anime(51), Anime(61)
+    )
+    every { refreshPage(any()) } returns Single.just(
+      Page(1, freshAnime.size, true, freshAnime, 0, 0)
+    )
+    presenter.bind(view)
+    presenter.bindIntents()
+    val expectedInitial = AnilistViewState(pageState = PageState(anime, 1, true))
+    val expectedLoading = AnilistViewState(
+      loading = REFRESH,
+      pageState = PageState(anime, 1, true),
+    )
+    val expectedResult = AnilistViewState(pageState = PageState(freshAnime, 1, true))
+    refreshIntentMock.onNext(Unit)
+    verifyOrder {
+      view.render(expectedInitial)
+      view.render(expectedLoading)
+      view.render(expectedResult)
+    }
+  }
+
+  @Test
+  fun `show details while refreshing test`() {
+    setupTrampoline()
+    val initialState = AnilistViewState(
+      loading = REFRESH,
+      pageState = PageState(anime, 1, true)
+    )
+    val presenter = AnilistPresenter(
+      loadPage, refreshPage, loadAnime, getStringResource, initialState, rxSchedulers
+    )
+    val showDetailsIntentMock: PublishSubject<Int> = PublishSubject.create()
+    mockIntents(showDetailsIntent = showDetailsIntentMock)
+    val animeDetails = Anime(
+      1,
+      "title",
+      Type.ANIME,
+      "test description",
+      FALL,
+      1998,
+      22,
+      17,
+      "image/test/url",
+      "banner/test/url",
+      0,
+      0
+    )
+    val expectedDetails = animeDetails.copy()
+    every { loadAnime(any()) } returns Single.just(animeDetails)
```

Comment:
Ahah, your expected state equals your tested state always
Something wrong here

Reply:
No, I test not details, but a state itself (which contains details). copying details prevents expected details change after test instance change in the tested class, so it is ok